rfcs/0049-flakes.md

```diff
@@ -0,0 +1,500 @@
+    not available for `github` repositories (see below), since they're
+    fetched as tarballs rather than as Git repositories.
+
+  * `lastModifiedDate`: The commit time of the revision `rev`, in the
+    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
+    this is available for both Git and GitHub repositories, so it's
+    useful for generating (hopefully) monotonically increasing version
+    strings.
+
+  * `lastModified`: The commit time of the revision `rev` as an integer
+    denoting the number of seconds since 1970.
+
+  * `narHash`: The SHA-256 (in SRI format) of the NAR serialization of
+    the flake's source tree.
+
+  The value returned by the `outputs` function must be an attribute
+  set. The attributes can have arbitrary values; however, some tools
+  may require specific attributes to have a specific value (e.g. the
+  `nix` command may expect the value of `packages.x86_64-linux` to be
+  an attribute set of derivations built for the `x86_64-linux`
+  platform).
+
+## Flake inputs
+
+The attribute `inputs` specifies the dependencies of a flake. These
+specify the location of the dependency, or a symbolic flake identifier
+that is looked up in a registry or in a command-line flag. For
+example, the following specifies a dependency on the Nixpkgs and Hydra
+repositories:
+
+    # A GitHub repository.
+    inputs.import-cargo = {
+      type = "github";
+      owner = "edolstra";
+      repo = "import-cargo";
+    };
+
+    # An indirection through the flake registry.
+    inputs.nixpkgs = {
+      type = "indirect";
+      id = "nixpkgs";
+    };
+
+Each input is fetched, evaluated and passed to the `outputs` function
+as a set of attributes with the same name as the corresponding
+input. The special input named `self` refers to the outputs and source
+tree of *this* flake. Thus, a typical `outputs` function looks like
+this:
+
+    outputs = { self, nixpkgs, import-cargo }: {
+      ... outputs ...
+    };
+
+It is also possible to omit inputs entirely and *only* list them as
+expected function arguments in `outputs`. Thus,
+
+    outputs = { self, nixpkgs }: ...;
+
+without an `inputs.nixpkgs` attribute will simply look up `nixpkgs` in
+the flake registry.
+
+Repositories that don't contain a `flake.nix` can also be used as
+inputs, by setting the input's `flake` attribute to `false`:
+
+    inputs.grcov = {
+      type = "github";
+      owner = "mozilla";
+      repo = "grcov";
+      flake = false;
+    };
+
+    outputs = { self, nixpkgs, grcov }: {
+      packages.x86_64-linux.grcov = stdenv.mkDerivation {
+        src = grcov;
+        ...
+      };
+    };
+
+The following input types are specified at present:
+
+* `git`: A Git repository or dirty local working tree.
+
+* `github`: A more efficient scheme to fetch repositories from GitHub
+  as tarballs. These have slightly different semantics from `git`
+  (in particular, the `revCount` attribute is not available).
```

Comment:
imo github option shouldn't be a part of the spec (but may have special shortcuts/tools in cli)
- tarball fetch is just an instance of http fetch
- the way github generates tarball urls may change in future
- for private repos user may have configured ssh access but not http (via tokens)

Reply:
It would be nice for things like github to be provided by external plugins, but github fetch *does* have different semantics than http fetch, in part because of the things you call out here! IMO it should be a separate primitive.

Reply:
Bitbucket for example also has tarball downloads of git repos, GitLab too, likely many others. So going this way would probably mean that in future more git (or even non-git) hosting services will need special handling.

Reply:
Other hosting services aren't the center of NixOS coordination, though..

Reply:
Right, most Nix-related repos are on GitHub, so in the interest of UX, it's nice to provide a shorter syntax and better semantics for GitHub. We don't have to support every other hosting service out there.

Reply:
I find that most repos use `feature/<name>` branches, but I didn't find a way to make a URI containing this branch name. Is this expected behavior?

Reply:
> Right, most Nix-related repos are on GitHub, so in the interest of UX, it's nice to provide a shorter syntax and better semantics for GitHub. We don't have to support every other hosting service out there.

@rycee uses many GitLab inputs for the `home-manager` repo. Is GitLab may be also desired?

Reply:
Yes, especially if we can get the `rev` attribute from GitLab.

Reply:
I create a Libfetcher for GitLab: https://github.com/NixOS/nix/pull/3636